src/matrix/room/timeline/persistence/GapWriter.js

```diff
@@ -252,3 +254,268 @@ export class GapWriter {
+        },
+        "Backfilling a fragment that is expected to link up, and does": async assert => {
+            const { txn, fragmentIdComparer, gapWriter, eventCreator } = await setup();
+            const existingFragment = await createFragment(0, txn, fragmentIdComparer, { nextId: 1, nextToken: startToken });
+            const liveFragment = await createFragment(1, txn, fragmentIdComparer, { previousId: 0, previousToken: startToken });
+            const newEntry = FragmentBoundaryEntry.start(liveFragment, fragmentIdComparer);
+
+            const initialEntries = await prefillFragment(txn, eventCreator, existingFragment, 10);
+            const response = eventCreator.createMessagesResponse();
+            response.chunk.push(initialEntries[initialEntries.length-1]); /* Expect overlap */
+            await gapWriter.writeFragmentFill(newEntry, response, txn, null);
+
+            const allEvents = await txn.timelineEvents._timelineStore.selectAll();
+            let i = 0;
+            for (const initialEntry of initialEntries) {
+                const storedEvent = allEvents[i++];
+                assert.deepEqual(initialEntry, storedEvent.event);
+            }
+            for (let j = 0; j < response.chunk.length - 1; j++) {
+                const responseEvent = response.chunk[response.chunk.length -j - 2];
+                const storedEvent = allEvents[i + j];
+                assert.deepEqual(responseEvent, storedEvent.event);
+            }
+            await assertTightLink(assert, txn, 0, 1);
+        },
+        "Backfilling a fragment that is expected to link up, but doesn't yet": async assert => {
+            const { txn, fragmentIdComparer, gapWriter, eventCreator } = await setup();
+            const existingFragment = await createFragment(0, txn, fragmentIdComparer, { nextId: 1, nextToken: endToken });
+            const liveFragment = await createFragment(1, txn, fragmentIdComparer, { previousId: 0, previousToken: startToken });
+            const newEntry = FragmentBoundaryEntry.start(liveFragment, fragmentIdComparer);
+
+            const initialEntries = await prefillFragment(txn, eventCreator, existingFragment, 10);
+            const response = eventCreator.createMessagesResponse();
+            await gapWriter.writeFragmentFill(newEntry, response, txn, null);
+
+            const allEvents = await txn.timelineEvents._timelineStore.selectAll();
+            let i = 0;
+            for (const initialEntry of initialEntries) {
+                const storedEvent = allEvents[i++];
+                assert.deepEqual(initialEntry, storedEvent.event);
+            }
+            for (let j = 0; j < response.chunk.length - 1; j++) {
+                const responseEvent = response.chunk[response.chunk.length - j - 1];
+                const storedEvent = allEvents[i + j];
+                assert.deepEqual(responseEvent, storedEvent.event);
+            }
+            await assertWeakLink(assert, txn, 0, 1);
```

Comment:
if we don't use static values for the pagination tokens, we can check the actual value of the token rather than it being truthy.